npm/vite-dev-server/src/resolveConfig.ts

```diff
@@ -79,28 +80,34 @@ function makeCypressViteConfig (config: ViteDevServerConfig, vite: Vite): Inline
     paths: [projectRoot],
   })))
 
+  const esbuildOptions: { incremental?: boolean, plugins: any[] } = {
+    incremental: true,
+    plugins: [
+      {
+        name: 'cypress-esbuild-plugin',
+        setup (build: any) {
+          build.onEnd(function (result: any) {
+            // We don't want to completely fail the build here on errors so we treat the errors as warnings
+            // which will handle things more gracefully. Vite will 500 on files that have errors when they
+            // are requested later and Cypress will display an error message.
+            // See: https://github.com/cypress-io/cypress/pull/21599
+            result.warnings = [...result.warnings, ...result.errors]
+            result.errors = []
+          })
+        },
+      },
+    ],
+  }
+
+  if (vite.version && semverGte(vite.version, '4.2.0')) {
+    delete esbuildOptions.incremental
+  }
+
```

Comment:
I think we can preserve the type safety a bit more, how about instead of add/delete, just add if needed:

```js
  const conf: InlineConfig = {
    root: projectRoot,
    base: `${devServerPublicPathRoute}/`,
    optimizeDeps: {
      esbuildOptions: {
        plugins: [
          {
            name: 'cypress-esbuild-plugin',
            setup (build) {
              build.onEnd(function (result) {
                // We don't want to completely fail the build here on errors so we treat the errors as warnings
                // which will handle things more gracefully. Vite will 500 on files that have errors when they
                // are requested later and Cypress will display an error message.
                // See: https://github.com/cypress-io/cypress/pull/21599
                result.warnings = [...result.warnings, ...result.errors]
                result.errors = []
              })
            },
          },
        ],
      },
      entries: [
        ...specs.map((s) => path.relative(projectRoot, s.relative)),
        ...(supportFile ? [path.resolve(projectRoot, supportFile)] : []),
      ].filter((v) => v != null),
    },
    // ... other stuff ...
   
  }

  // add it here
  if (vite_version > 4.2.0 ^^ conf?.optimizeDeps?.esbuildOptions) {
    conf.optimizeDeps.esbuildOptions.incremental = true
  }

  return conf
```

Reply:
Good idea, thanks